HIP/hip-968.md

```diff
@@ -0,0 +1,131 @@
+---
+hip: 968
+title: Enhance mirror node rest API to query by token name
+author: Simi Hunjan <SimiHunjan>, Ali Nikan <@alinik4n>
+working-group: Steven Sheehy <@steven-sheehy>, Eric <ericleponner>, Simon <svienot>
+type: Standards Track
+category: Mirror
+needs-council-approval: Yes
+status: Review
+created: 2024-04-19
+updated: 2024-05-14
+requested-by: Hashport, Bonzo Finance
+---
+
+## Abstract
+
+This HIP provides the capability of querying tokens by token name in mirror node REST API and getting the corresponding token ID information.
+
+## Motivation
+
+To enhance user experience and allow blockchain explorers such as Hashscan to enable the capability of searching by token name, we want to add this functionality to the mirror node to support querying tokens by token name.
+This is a very common use case across other explorers in the larger ecosystem, and will allow retail users and developers to be able to search for a token by token name instead of token ID.
+
+## Rationale
+
+Many times a user may not know the exact token ID when searching for a token, but they know the name. This capability allow users to search by the name and will return results when there is a partial match or exact match. Community members from bridges, Dexes, etc. have been requesting for this for a while.
+
+## User stories
+
+- As a user of mirror node, I want to be able to query with a token name (e.g Sauce) and get the corresponding token ID and information.
+- As a user of mirror node, I want to be able to see full match and partial match to my search. (e.g when querying for Sauce, I want to see result for $sauce and $xsauce)
+- As a user of a Hedera explorer, I want to look up a token by searching for token name.
+
+## Specification
+
+The proposed enhancement involves adding  query parameters `name` to the existing route `/api/v1/tokens` to return information about a token.
+
+**Multiple Query Results**
+
+Names assigned to tokens when tokens are created on the network are not unique. Therefore, many tokens with the same name can exist. Querying a token by a given name will return all the tokens that have the specified name defined in the token name.
+
+**Name Query Parameter**
+
+It enables to specify value that token names should match. The following form can be specified:
+
+- name={value}
+- name=`lk`:{value} where `lk` is a new operator
+
+`name`=`lk`:{value} returns all tokens whose name:
+
+- starts with {value}
+- ends with {value}
+- contains {value}
+
+Comparison performed by `lk` operator should be case insensitive.
```

Comment:
We're not supporting a new `lk` operator. Instead, only `name={value}` will be supported and this allows case insensitive sub-string matches.

```suggestion
A new `name` query parameter will be added to search for tokens by their name. This query parameter will perform a case insensitive sub-string match. That is, it will search for token names that start with, contain, or end with the given value. The usual operators like `eq`, `gt`, `gte`, `lt, `lte`, and `ne` will not be supported.

`/api/v1/tokens?name={value}`
```